### PR TITLE
Added tests to cover all branches of JsonNodeOptions

### PR DESCRIPTION
**Pull request by AntonLandor**
Added tests to cover all branches of JsonNodeOptions. I am unsure whether JsonNode.TraversalTests was the right place for this test case, but I found no other more suitable place for it, so feedback is welcome. 

Solves [dotnet/corefx #42585](https://github.com/dotnet/corefx/issues/42585).

### CODE REVIEW

File: src/libraries/System.Text.Json/tests/JsonNode.TraversalTests.cs [hunk 1]
```diff
@@ -279,5 +279,26 @@ void CheckNestedValues(JsonObject jsonObject)
             Assert.IsType<JsonArray>(jsonObject["array"]);
             Assert.IsType<JsonNumber>(nestedObject["array"]);
         }
+
+        [Fact]
+        public static void TestInvalidJsonNodeOptions()
```
Comment (ahsonkhan) [inline src/libraries/System.Text.Json/tests/JsonNode.TraversalTests.cs]: Can you also add a test for what the values of a default `JsonNodeOptions` are?

So, assert the values of the options and that `JsonNodeOptions options = default;` and `var options = new JsonNodeOptions()` are equal for each of the options.

**Comment by ahsonkhan** [comment]
>  I am unsure whether JsonNode.TraversalTests was the right place for this test case, but I found no other more suitable place for it, so feedback is welcome.

Given there aren't a lot of tests related to `JsonNodeOptions`, I would be fine with leaving them where they are.

The other option would be to move them to a new `JsonNodeOptionsTests.cs` file, which I would be OK with as well.

Other tests we likely want to add, is setting various `JsonNodeOptions`, pass them to `JsonNode.Parse` and verify the behavior of `JsonNode` meets the expectation. For example, set max depth to 5 and verify that we throw for an input that has higher depth. Similarly for comments. The `DuplicatePropertyNameHandling` option is already sufficiently covered. 

If we are adding those new tests, a dedicated `JsonNodeOptionsTests.cs` file would be better.

File: src/libraries/System.Text.Json/tests/JsonNode.TraversalTests.cs (hunk 1 above)
Comment (AntonLandor) [reply src/libraries/System.Text.Json/tests/JsonNode.TraversalTests.cs]: Absolutely. I'll add it soon.

**Comment by AntonLandor** [comment]
> > I am unsure whether JsonNode.TraversalTests was the right place for this test case, but I found no other more suitable place for it, so feedback is welcome.
> 
> Given there aren't a lot of tests related to `JsonNodeOptions`, I would be fine with leaving them where they are.
> 
> The other option would be to move them to a new `JsonNodeOptionsTests.cs` file, which I would be OK with as well.
> 
> Other tests we likely want to add, is setting various `JsonNodeOptions`, pass them to `JsonNode.Parse` and verify the behavior of `JsonNode` meets the expectation. For example, set max depth to 5 and verify that we throw for an input that has higher depth. Similarly for comments. The `DuplicatePropertyNameHandling` option is already sufficiently covered.

The example of exceeding a max depth is already implemented in [`JsonNode.TraversalTests.cs`](https://github.com/dotnet/runtime/blob/21dc0db53b16f0c8ce99bde87363392950d9a4f3/src/libraries/System.Text.Json/tests/JsonNode.TraversalTests.cs#L112-L127) which in my opinion fits well in `JsonNode.TraversalTests.cs`, but also the the proposed new file, `JsonNodeOptionsTests.cs`. If we want to create these new tests I could move this test and others I find fitting to `JsonNodeOptionsTests.cs`. 

> 
> If we are adding those new tests, a dedicated `JsonNodeOptionsTests.cs` file would be better.

I agree though I imagine most if not all the proposed tests could be placed in `JsonNode.TraversalTests.cs` since they are passed to `JsonNode.Parse`, `JsonNodeOptionsTests.cs` would be better since we are specifically testing that setting `JsonNodeOptions` meets the expected behavior.

I could try to implement some new tests for `JsonNodeOptionsTests.cs` if you think it is a good idea.

**Comment by ahsonkhan** [comment]
> The example of exceeding a max depth is already implemented in [`JsonNode.TraversalTests.cs`](https://github.com/dotnet/runtime/blob/21dc0db53b16f0c8ce99bde87363392950d9a4f3/src/libraries/System.Text.Json/tests/JsonNode.TraversalTests.cs#L112-L127)

That particular test isn't customizing the `MaxDepth` in the options. Just relying on the default (which is 64). And then the other one (`TestParseDoesNotStackOverflow`) is about stackoverflow and also isn't setting a depth less than the payload and verifying we are throwing once `MaxDepth` is overriden.

> but also the the proposed new file, `JsonNodeOptionsTests.cs`. If we want to create these new tests I could move this test and others I find fitting to `JsonNodeOptionsTests.cs`. 

That works for me. I don't see `TestParseFailsWhenExceedsMaxDepth` being strongly related to Traversal, and it makes sense to move it to `JsonNodeOptionsTests.cs`.

> I could try to implement some new tests for `JsonNodeOptionsTests.cs` if you think it is a good idea @ahsonkhan.

Perfect :)

File: src/libraries/System.Text.Json/tests/System.Text.Json.Tests.csproj [hunk 2]
```diff
@@ -20,6 +20,7 @@
     <Compile Include="JsonElementWriteTests.cs" />
     <Compile Include="JsonEncodedTextTests.cs" />
     <Compile Include="JsonGuidTestData.cs" />
+    <Compile Include="JsonNodeOptionsTests.cs" />
```
Comment (ahsonkhan) [inline src/libraries/System.Text.Json/tests/System.Text.Json.Tests.csproj]: nit: Move this to the other ItemGroup which contains all the `JsonNode` related test files. We are keeping them separate until the API shape has been reviewed/finalized to make it easier to track/update.

https://github.com/dotnet/runtime/blob/f3be2149c44400a3fefbc7055240a6a9169c58cd/src/libraries/System.Text.Json/tests/System.Text.Json.Tests.csproj#L133

File: src/libraries/System.Text.Json/tests/JsonNodeOptionsTests.cs [hunk 3]
```diff
@@ -0,0 +1,107 @@
+﻿// Licensed to the .NET Foundation under one or more agreements.
+// The .NET Foundation licenses this file to you under the MIT license.
+// See the LICENSE file in the project root for more information.
+
+using Xunit;
+
+namespace System.Text.Json.Tests
+{
+    public static partial class JsonNodeOptionsTests
```
Comment (ahsonkhan) [inline src/libraries/System.Text.Json/tests/JsonNodeOptionsTests.cs]: Add a default JsonNodeOptions test.

**Review by ahsonkhan — APPROVED** [review]
Otherwise, look's good.

**Comment by ahsonkhan** [comment]
Unrelated `System.Security.Cryptography.OpenSsl.Tests` failure: https://github.com/dotnet/corefx/issues/38498

**Comment by MattGal** [comment]
Filed https://github.com/dotnet/core-eng/issues/8481 ; that machine got reimaged yesterday and has failed that work item 100% of times it's been given it.  I don't see what's wrong with the machine, so I've also reached out to @bartonjs to see if we can investigate the machine; whatever's wrong with it is not obvious as all the files for OpenSSL are where I'd expect them.

File: src/libraries/System.Text.Json/tests/System.Text.Json.Tests.csproj (hunk 2 above)
Comment (AntonLandor) [reply src/libraries/System.Text.Json/tests/System.Text.Json.Tests.csproj]: Done. I will fix the other suggestion soon.

File: src/libraries/System.Text.Json/tests/JsonNodeOptionsTests.cs (hunk 3 above)
Comment (AntonLandor) [reply src/libraries/System.Text.Json/tests/JsonNodeOptionsTests.cs]: Done. I also added tests using default values with JsoneNode.Parse(...).

File: src/libraries/System.Text.Json/tests/JsonNodeOptionsTests.cs [hunk 4]
```diff
@@ -0,0 +1,120 @@
...
+            });
+        }
+
+        [Fact]
+        public static void TestDefaultJsonNodeOptions()
+        {
+            JsonNodeOptions defaultOptions = default;
+            JsonNodeOptions newOptions = new JsonNodeOptions();
+
+            Assert.Equal(defaultOptions.AllowTrailingCommas, newOptions.AllowTrailingCommas);
```
Comment (ahsonkhan) [inline src/libraries/System.Text.Json/tests/JsonNodeOptionsTests.cs]: nit: Can you also add asserts for what the `defaultOptions` are?

```C#
Assert.Equals(false, defaultOptions.AllowTrailingCommas);
Assert.Equals(JsonCommentHandling.Disallow, defaultOptions.CommentHandling);
Assert.Equals(DuplicatePropertyNameHandlingStrategy.Replace, defaultOptions.DuplicatePropertyNameHandling);
Assert.Equals(0, defaultOptions.MaxDepth);
```

**Review by ahsonkhan — APPROVED** [review]
Otherwise, looks good.

**Comment by ahsonkhan** [comment]
@dotnet/dnceng This looks like an infra related error - operation canceled after some retries (in the `Send to Helix` phase - for the `runtime-libraries (Libraries Build Windows_NT x86 Release)` CI leg).

https://dev.azure.com/dnceng/public/_build/results?buildId=451661

```text

Build FAILED.

F:\workspace\_work\1\s\.packages\microsoft.dotnet.helix.sdk\5.0.0-beta.19604.4\tools\Microsoft.DotNet.Helix.Sdk.MonoQueue.targets(47,5): error : AggregateException: Retry failed after 4 tries. (The operation was canceled.) (The operation was canceled.) (The operation was canceled.) (The operation was canceled.) [F:\workspace\_work\1\s\src\libraries\sendtohelix.proj]
F:\workspace\_work\1\s\.packages\microsoft.dotnet.helix.sdk\5.0.0-beta.19604.4\tools\Microsoft.DotNet.Helix.Sdk.MonoQueue.targets(47,5): error :    at Azure.Core.Pipeline.RetryPolicy.ProcessAsync(HttpMessage message, ReadOnlyMemory`1 pipeline, Boolean async) [F:\workspace\_work\1\s\src\libraries\sendtohelix.proj]
F:\workspace\_work\1\s\.packages\microsoft.dotnet.helix.sdk\5.0.0-beta.19604.4\tools\Microsoft.DotNet.Helix.Sdk.MonoQueue.targets(47,5): error :    at Azure.Core.Pipeline.HttpPipelineSynchronousPolicy.ProcessAsync(HttpMessage message, ReadOnlyMemory`1 pipeline) [F:\workspace\_work\1\s\src\libraries\sendtohelix.proj]
F:\workspace\_work\1\s\.packages\microsoft.dotnet.helix.sdk\5.0.0-beta.19604.4\tools\Microsoft.DotNet.Helix.Sdk.MonoQueue.targets(47,5): error :    at Azure.Core.Pipeline.HttpPipelineSynchronousPolicy.ProcessAsync(HttpMessage message, ReadOnlyMemory`1 pipeline) [F:\workspace\_work\1\s\src\libraries\sendtohelix.proj]
F:\workspace\_work\1\s\.packages\microsoft.dotnet.helix.sdk\5.0.0-beta.19604.4\tools\Microsoft.DotNet.Helix.Sdk.MonoQueue.targets(47,5): error :    at Azure.Core.Pipeline.HttpPipeline.SendRequestAsync(Request request, CancellationToken cancellationToken) [F:\workspace\_work\1\s\src\libraries\sendtohelix.proj]
F:\workspace\_work\1\s\.packages\microsoft.dotnet.helix.sdk\5.0.0-beta.19604.4\tools\Microsoft.DotNet.Helix.Sdk.MonoQueue.targets(47,5): error :    at Microsoft.DotNet.Helix.Client.Job.NewAsync(JobCreationRequest body, String idempotencyKey, CancellationToken cancellationToken) in /_/src/Microsoft.DotNet.Helix/Client/CSharp/generated-code/Job.cs:line 212 [F:\workspace\_work\1\s\src\libraries\sendtohelix.proj]
F:\workspace\_work\1\s\.packages\microsoft.dotnet.helix.sdk\5.0.0-beta.19604.4\tools\Microsoft.DotNet.Helix.Sdk.MonoQueue.targets(47,5): error :    at Microsoft.DotNet.Helix.Client.JobDefinition.SendAsync(Action`1 log, CancellationToken cancellationToken) in /_/src/Microsoft.DotNet.Helix/JobSender/JobDefinition.cs:line 219 [F:\workspace\_work\1\s\src\libraries\sendtohelix.proj]
F:\workspace\_work\1\s\.packages\microsoft.dotnet.helix.sdk\5.0.0-beta.19604.4\tools\Microsoft.DotNet.Helix.Sdk.MonoQueue.targets(47,5): error :    at Microsoft.DotNet.Helix.Sdk.SendHelixJob.ExecuteCore(CancellationToken cancellationToken) in /_/src/Microsoft.DotNet.Helix/Sdk/SendHelixJob.cs:line 218 [F:\workspace\_work\1\s\src\libraries\sendtohelix.proj]
F:\workspace\_work\1\s\.packages\microsoft.dotnet.helix.sdk\5.0.0-beta.19604.4\tools\Microsoft.DotNet.Helix.Sdk.MonoQueue.targets(47,5): error :    at Microsoft.DotNet.Helix.Sdk.HelixTask.Execute() in /_/src/Microsoft.DotNet.Helix/Sdk/HelixTask.cs:line 51 [F:\workspace\_work\1\s\src\libraries\sendtohelix.proj]
F:\workspace\_work\1\s\.packages\microsoft.dotnet.helix.sdk\5.0.0-beta.19604.4\tools\Microsoft.DotNet.Helix.Sdk.MonoQueue.targets(47,5): error : TaskCanceledException: The operation was canceled. [F:\workspace\_work\1\s\src\libraries\sendtohelix.proj]
F:\workspace\_work\1\s\.packages\microsoft.dotnet.helix.sdk\5.0.0-beta.19604.4\tools\Microsoft.DotNet.Helix.Sdk.MonoQueue.targets(47,5): error :    at System.Net.Http.HttpConnection.SendAsyncCore(HttpRequestMessage request, CancellationToken cancellationToken) [F:\workspace\_work\1\s\src\libraries\sendtohelix.proj]
F:\workspace\_work\1\s\.packages\microsoft.dotnet.helix.sdk\5.0.0-beta.19604.4\tools\Microsoft.DotNet.Helix.Sdk.MonoQueue.targets(47,5): error :    at System.Net.Http.HttpConnectionPool.SendWithNtConnectionAuthAsync(HttpConnection connection, HttpRequestMessage request, Boolean doRequestAuth, CancellationToken cancellationToken) [F:\workspace\_work\1\s\src\libraries\sendtohelix.proj]
F:\workspace\_work\1\s\.packages\microsoft.dotnet.helix.sdk\5.0.0-beta.19604.4\tools\Microsoft.DotNet.Helix.Sdk.MonoQueue.targets(47,5): error :    at System.Net.Http.HttpConnectionPool.SendWithRetryAsync(HttpRequestMessage request, Boolean doRequestAuth, CancellationToken cancellationToken) [F:\workspace\_work\1\s\src\libraries\sendtohelix.proj]
F:\workspace\_work\1\s\.packages\microsoft.dotnet.helix.sdk\5.0.0-beta.19604.4\tools\Microsoft.DotNet.Helix.Sdk.MonoQueue.targets(47,5): error :    at System.Net.Http.RedirectHandler.SendAsync(HttpRequestMessage request, CancellationToken cancellationToken) [F:\workspace\_work\1\s\src\libraries\sendtohelix.proj]
F:\workspace\_work\1\s\.packages\microsoft.dotnet.helix.sdk\5.0.0-beta.19604.4\tools\Microsoft.DotNet.Helix.Sdk.MonoQueue.targets(47,5): error :    at System.Net.Http.HttpClient.FinishSendAsyncUnbuffered(Task`1 sendTask, HttpRequestMessage request, CancellationTokenSource cts, Boolean disposeCts) [F:\workspace\_work\1\s\src\libraries\sendtohelix.proj]
F:\workspace\_work\1\s\.packages\microsoft.dotnet.helix.sdk\5.0.0-beta.19604.4\tools\Microsoft.DotNet.Helix.Sdk.MonoQueue.targets(47,5): error :    at Azure.Core.Pipeline.HttpClientTransport.ProcessAsync(HttpMessage message) [F:\workspace\_work\1\s\src\libraries\sendtohelix.proj]
F:\workspace\_work\1\s\.packages\microsoft.dotnet.helix.sdk\5.0.0-beta.19604.4\tools\Microsoft.DotNet.Helix.Sdk.MonoQueue.targets(47,5): error :    at Azure.Core.Pipeline.RequestActivityPolicy.ProcessNextAsync(HttpMessage message, ReadOnlyMemory`1 pipeline, Boolean isAsync) [F:\workspace\_work\1\s\src\libraries\sendtohelix.proj]
F:\workspace\_work\1\s\.packages\microsoft.dotnet.helix.sdk\5.0.0-beta.19604.4\tools\Microsoft.DotNet.Helix.Sdk.MonoQueue.targets(47,5): error :    at Azure.Core.Pipeline.RequestActivityPolicy.ProcessAsync(HttpMessage message, ReadOnlyMemory`1 pipeline, Boolean isAsync) [F:\workspace\_work\1\s\src\libraries\sendtohelix.proj]
F:\workspace\_work\1\s\.packages\microsoft.dotnet.helix.sdk\5.0.0-beta.19604.4\tools\Microsoft.DotNet.Helix.Sdk.MonoQueue.targets(47,5): error :    at Azure.Core.Pipeline.BufferResponsePolicy.ProcessAsync(HttpMessage message, ReadOnlyMemory`1 pipeline) [F:\workspace\_work\1\s\src\libraries\sendtohelix.proj]
F:\workspace\_work\1\s\.packages\microsoft.dotnet.helix.sdk\5.0.0-beta.19604.4\tools\Microsoft.DotNet.Helix.Sdk.MonoQueue.targets(47,5): error :    at Azure.Core.Pipeline.LoggingPolicy.ProcessAsync(HttpMessage message, ReadOnlyMemory`1 pipeline, Boolean async) [F:\workspace\_work\1\s\src\libraries\sendtohelix.proj]
F:\workspace\_work\1\s\.packages\microsoft.dotnet.helix.sdk\5.0.0-beta.19604.4\tools\Microsoft.DotNet.Helix.Sdk.MonoQueue.targets(47,5): error :    at Azure.Core.Pipeline.LoggingPolicy.ProcessAsync(HttpMessage message, ReadOnlyMemory`1 pipeline) [F:\workspace\_work\1\s\src\libraries\sendtohelix.proj]
F:\workspace\_work\1\s\.packages\microsoft.dotnet.helix.sdk\5.0.0-beta.19604.4\tools\Microsoft.DotNet.Helix.Sdk.MonoQueue.targets(47,5): error :    at Azure.Core.Pipeline.RetryPolicy.ProcessAsync(HttpMessage message, ReadOnlyMemory`1 pipeline, Boolean async) [F:\workspace\_work\1\s\src\libraries\sendtohelix.proj]
F:\workspace\_work\1\s\.packages\microsoft.dotnet.helix.sdk\5.0.0-beta.19604.4\tools\Microsoft.DotNet.Helix.Sdk.MonoQueue.targets(47,5): error : IOException: Unable to read data from the transport connection: The I/O operation has been aborted because of either a thread exit or an application request.. [F:\workspace\_work\1\s\src\libraries\sendtohelix.proj]
F:\workspace\_work\1\s\.packages\microsoft.dotnet.helix.sdk\5.0.0-beta.19604.4\tools\Microsoft.DotNet.Helix.Sdk.MonoQueue.targets(47,5): error :    at System.Net.Sockets.Socket.AwaitableSocketAsyncEventArgs.ThrowException(SocketError error, CancellationToken cancellationToken) [F:\workspace\_work\1\s\src\libraries\sendtohelix.proj]
F:\workspace\_work\1\s\.packages\microsoft.dotnet.helix.sdk\5.0.0-beta.19604.4\tools\Microsoft.DotNet.Helix.Sdk.MonoQueue.targets(47,5): error :    at System.Net.Sockets.Socket.AwaitableSocketAsyncEventArgs.GetResult(Int16 token) [F:\workspace\_work\1\s\src\libraries\sendtohelix.proj]
F:\workspace\_work\1\s\.packages\microsoft.dotnet.helix.sdk\5.0.0-beta.19604.4\tools\Microsoft.DotNet.Helix.Sdk.MonoQueue.targets(47,5): error :    at System.Net.Security.SslStream.<FillBufferAsync>g__InternalFillBufferAsync|217_0[TReadAdapter](TReadAdapter adap, ValueTask`1 task, Int32 min, Int32 initial) [F:\workspace\_work\1\s\src\libraries\sendtohelix.proj]
F:\workspace\_work\1\s\.packages\microsoft.dotnet.helix.sdk\5.0.0-beta.19604.4\tools\Microsoft.DotNet.Helix.Sdk.MonoQueue.targets(47,5): error :    at System.Net.Security.SslStream.ReadAsyncInternal[TReadAdapter](TReadAdapter adapter, Memory`1 buffer) [F:\workspace\_work\1\s\src\libraries\sendtohelix.proj]
F:\workspace\_work\1\s\.packages\microsoft.dotnet.helix.sdk\5.0.0-beta.19604.4\tools\Microsoft.DotNet.Helix.Sdk.MonoQueue.targets(47,5): error :    at System.Net.Http.HttpConnection.SendAsyncCore(HttpRequestMessage request, CancellationToken cancellationToken) [F:\workspace\_work\1\s\src\libraries\sendtohelix.proj]
F:\workspace\_work\1\s\.packages\microsoft.dotnet.helix.sdk\5.0.0-beta.19604.4\tools\Microsoft.DotNet.Helix.Sdk.MonoQueue.targets(47,5): error : SocketException: The I/O operation has been aborted because of either a thread exit or an application request. [F:\workspace\_work\1\s\src\libraries\sendtohelix.proj]
F:\workspace\_work\1\s\.packages\microsoft.dotnet.helix.sdk\5.0.0-beta.19604.4\tools\Microsoft.DotNet.Helix.Sdk.MonoQueue.targets(47,5): error :  [F:\workspace\_work\1\s\src\libraries\sendtohelix.proj]
F:\workspace\_work\1\s\.packages\microsoft.dotnet.helix.sdk\5.0.0-beta.19604.4\tools\Microsoft.DotNet.Helix.Sdk.MonoQueue.targets(47,5): error :  [F:\workspace\_work\1\s\src\libraries\sendtohelix.proj]
    0 Warning(s)
    1 Error(s)
```

**Comment by riarenas** [comment]
@ahsonkhan I filed https://github.com/dotnet/core-eng/issues/8486 to take a look at this. I'd recommend retrying, but with the current AzDO problems when queuing builds...

**Comment by MattGal** [comment]
> @ahsonkhan I filed [dotnet/core-eng#8486](https://github.com/dotnet/core-eng/issues/8486) to take a look at this. I'd recommend retrying, but with the current AzDO problems when queuing builds...

Retry all you like, but odds are nothing will happen (I'm on the bridge right now).  I'll take a peek while I sit here listening.

File: src/libraries/System.Text.Json/tests/JsonNodeOptionsTests.cs (hunk 4 above)
Comment (AntonLandor) [reply src/libraries/System.Text.Json/tests/JsonNodeOptionsTests.cs]: Done!

**Pull request merged**